Rework expression evaluation.

petermattis opened the pull request:
Removed support for unsigned arithmetic. Removed much, but not all, of
the implicit casting. Switched to using table-based dispatch for unary,
binary and comparison operators. Fixed three-valued boolean logic now
that we can parse such expressions correctly.

petermattis committed:
Replace sqlwire.Datum usage with parser2.Datum.

This makes some of the code more idiomatic go and paves the way for
supporting row operations ([]Datum).

petermattis committed:
Switch to table-based lookup of unary, binary and comparison ops.

petermattis committed:
Remove unnecessary conversions from string to {bool,int,float}.

petermattis committed:
Fix three-valued boolean logic.

Fixes #1621.

petermattis committed:
Remove unnecessary conversion routines.

tamird commented on sql/parser2/eval.go:
```diff
@@ -20,31 +20,308 @@ package parser2
+// A Datum holds either a bool, int64, float64, string or []Datum.
+type Datum interface {
+	IsNull() bool
+	ToBool() (Datum, error)
```
looks like all calls to this are followed by a type assertion to `dbool`. can you make this just return `dbool`?

petermattis replied:
Yeah, this code evolved a bit. Let me try that out, sounds simpler.

petermattis replied:
Done. PTAL.

tamird commented on sql/parser2/eval.go:
```diff
@@ -20,31 +20,308 @@ package parser2
+	return "NULL"
+}
+
+var null = dnull{}
```
const?

petermattis replied:
Compiler complains that `const initializer dnull literal is not a constant`. I think constants can only be basic types.

tamird commented on sql/parser2/eval.go:
```diff
@@ -20,31 +20,308 @@ package parser2
+
+// binOps contains the binary operations indexed by operation type and argument
+// types.
+var binOps = map[binArgs]func(Datum, Datum) (Datum, error){
```
can you extract the function signature into a local type? I think it'd make these easier to read and review, especially since none of them seems to return non-nil errors

petermattis replied:
I'm not sure what you're after. I can define type for binary funcs:

```
  type binFunc func(Datum, Datum) (Datum, error)
```

But when defining the values for the map I have to spell out the function signature. So the only place I can use the type is in the map definition. Am I missing something.

Yes, all of these methods currently return `nil` for an error, but at some point soon I need to check for overflow/underflow/divide-by-zero/etc.

tamird replied:
> But when defining the values for the map I have to spell out the function signature.

Right, disregard.

tamird commented on sql/parser2/eval.go:
```diff
@@ -284,174 +536,46 @@ func evalComparisonExpr(expr *ComparisonExpr, env Env) (sqlwire.Datum, error) {
-		default:
-			panic(fmt.Sprintf("unsupported op type: %d", typ))
+	f := cmpOps[cmpArgs{op, reflect.TypeOf(left), reflect.TypeOf(right)}]
+	if f != nil {
```
should this panic if `f` is nil?

tamird replied:
ah, nevermind

petermattis replied:
Yep, if the function cannot be found the desired operation does not exist for the argument types. For example, you can't compare an int with a string.

petermattis committed:
Replace Datum.IsNull() with comparisons to null.

Renamed Datum.ToBool() to Datum.Bool() and change the return value to
dbool as that was always how the value was being used.

Added Datum.Type() to return the type name of the datum ("bool", "int",
etc).

tschottdorf commented on sql/parser2/eval.go:
```diff
@@ -20,31 +20,310 @@ package parser2
-	uintOp
-	floatOp
-	stringOp
+// TODO(pmattis): Do we want to allow implicit conversions of int to bool?
```
s/int/float/

petermattis replied:
Done. Also referenced #1626 in the comment.

tschottdorf commented:
LGTM. A lot more DRY than before.

tamird commented:
LGTM
